### PR TITLE
b64ct: fix overflow bug, add alloc-gated tests

### LINKED ISSUE #205: b64ct: subtle overflow bug
While working on the in-place decoding function, I noticed a subtle overflow bug. Currently `decoded_len` is implemented as:
```rust
pub const fn decoded_len(bytes: &str) -> usize {
    (bytes.len() * 3) / 4
}
```
For lengths greater than `usize::MAX/3` it will produce incorrect results. Obviously in practice it's really unlikely that someone will work with more than gigabyte B64-encoded strings and on 64-bit platforms it will be practically impossible to trigger this bug. But nevertheless it's a clear bug.

The easiest solution would be to cast `usize` to `u128` and back, but it's quite inelegant solution and it looks like a bit hard for compiler to properly infer properties of a resulting value... Another solution would be to use two branches for values bigger than `usize::MAX/3` and not. And the last solution will be to use an alternative formula which would produce identical results.

The same applies to `encoded_len` as well, since it contains `bytes.len() * 4`.

### PR DESCRIPTION
**Pull request by newpavlov**
Closes #205

### CODE REVIEW

File: b64ct/src/lib.rs [hunk 1]
```diff
@@ -70,10 +73,13 @@ pub fn encode<'a>(src: &[u8], dst: &'a mut [u8]) -> Result<&'a str, InvalidLengt
...
 
 /// Encode the input byte slice as a "B64"-encoded [`String`].
+///
+/// # Panics
+/// If `input` length is greater than `usize::MAX/4`.
 #[cfg(feature = "alloc")]
 #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
 pub fn encode_string(input: &[u8]) -> String {
-    let elen = encoded_len(input);
+    let elen = encoded_len_inner(input.len()).expect("input is too big");
```
Comment (tarcieri) [inline b64ct/src/lib.rs:82]: I'd be ok with making this API fallible for this case

**Review by tarcieri — APPROVED** [review]
Great, thank you!

File: b64ct/src/lib.rs (hunk 1 above)
Comment (newpavlov) [reply b64ct/src/lib.rs:82]: I am not sure if it's worth the trouble. In the first place it's a convenience method and probability of encountering values larger than 1 GiB is extremely small. If someone wants to be strictly panic-free, they always can use the underlying function directly.

File: b64ct/src/lib.rs [hunk 2]
```diff
@@ -200,7 +221,18 @@ pub const fn decoded_len(bytes: &str) -> usize {
...
+    type DoubleWord = u128;
+    #[cfg(target_pointer_width = "32")]
+    type DoubleWord = u64;
+    #[cfg(target_pointer_width = "16")]
+    type DoubleWord = u32;
+
+    if n <= usize::MAX / 3 {
+        (3 * n) / 4
+    } else {
+        ((3 * (n as DoubleWord)) / 4) as usize
```
Comment (newpavlov) [inline b64ct/src/lib.rs:234]: I think it should be possible to do it in a branch-less fashion without introducing the `DoubleWord` type...

**Comment by newpavlov** [comment]
Matches and if branches in `const fn`s have MSRV of 1.46. What do you think is worth doing here? Bump crate MSRV, make the functions non-const, try to find branch-less tricks?

**Comment by tarcieri** [comment]
MSRV 1.46+ is fine by me!

All of the dependent crates which use `b64ct` (i.e. `password-hash` and `pkcs8`) are MSRV 1.46+ anyway.

**Pull request merged**